tests/python_validation/python_validate.py

```diff
@@ -0,0 +1,205 @@
+        for k, line in enumerate(lines):
+            print(("%{0}d:".format(nchar))%(k+1), repr(line)[1:-1])
+    return fail
+
+def validate_file(filename, fix):
+    """Run the validotor on a whole file."""
+    with open(filename, 'r') as testfile:
+        failed = process_snippet(filename, testfile.read())
+        if failed:
+            if fix:
+                print_as_function(filename)
+            return True
+    return False
+
+
+def validate_xml_snippets(filename, test_pep8, fix):
+    """Run the validotor on elements with attribute 'language=python' in an xml file."""
+
+    snippets=[]
+
+    with open(filename, 'r') as xmlfile:
+        tree = etree.ElementTree(file=filename)
+
+
+        for ele in tree.iter():
+            if ele.attrib.get('language',None)=='python':
+                snippets.append((get_name(ele), ele.text, ele))
+
+        failed = False
+        modified = False
+        for name, snippet, ele in snippets:
+            if(test_pep8):
+                flake8_snippet(name, snippet)
+            if process_snippet(name, snippet):
+                failed = True
+                if fix:
+                    fixed = fix_snippet(name, snippet)
+                    if fixed:
+                        modified = True
+                        ele.text = fixed
+                else:
+                    return True
+        if failed and modified:
+            tree.write(filename, xml_declaration=True, encoding='utf-8',
+                   with_tail=False, pretty_print=True)
+
+    return failed
+
+def main():
+    """Main python validation routine."""
+
+    parser = argparse.ArgumentParser(usage="<python_executable> python_validate.py [-p] [-f] [file1 file2 file3]")
+    parser.add_argument("-f", "--fixup",  action="store_true", dest="fix", default=False,
+                  help="Automatically attempt to fix obvious python3 errors using reindent and 2to3")
+    parser.add_argument("-p", "--pep8", action="store_true", dest="test_pep8", default=False,
+                  help="Run flake8 to test adherence to pep8 coding stanaards.")
+    parser.add_argument('files', nargs='*',
+                    help='files to process')
+    options = parser.parse_args()
+    failures = []
+
+    failed = False
+    print(options.files)
+    for filename in options.files:
+        if filename.rsplit('.',1)[-1]=='py':
+            if validate_file(filename, options.fix):
+                failed =True
+                failures.append(filename)
+        else:
+            if validate_xml_snippets(filename, 
+                                     test_pep8=options.test_pep8,
+                                     fix=options.fix):
+                failed =True
+                failures.append(filename)
+
+    if failed:
+        print("Failed:")
+        for filename in failures:
+            print(filename)
+        exit(1)
+    else:
+        print("Passed!")
+        exit(0)
+
+def check_versions():
```

Comment:
Is this actually used?

Reply:
My intention was that check_versions be called as a warn_test to check python2 and python3 both exist on testing machines, with the validation then proceeding for all the versions which exist. The idea being that we can't expect people to validate a python version they don't have installed. The buildbot results show that isn't precisely happening as well as I'd hoped.